File: server/src/routes/sessions.ts
import { Router } from 'express';
import { unlinkSync } from 'fs';
import {
  listSessions,
  getSession,
  getSessionMessages,
  getSessionStats,
  findSessionFiles,
} from '../services/sessionService.js';
import type { SessionSource } from '../parsers/index.js';

export const sessionsRouter = Router();

// GET /api/sessions - List all sessions
sessionsRouter.get('/', (req, res) => {
  try {
    const source = req.query.source as 'claude' | 'copilot' | 'all' | undefined;
    const sessions = listSessions(source);
    res.json(sessions);
  } catch (error) {
    console.error('Error listing sessions:', error);
    res.status(500).json({ error: 'Internal server error', message: 'Failed to list sessions' });
  }
});

// GET /api/sessions/:source/:sessionId - Get session details
sessionsRouter.get('/:source/:sessionId', (req, res) => {
  try {
    const { source, sessionId } = req.params;

    if (source !== 'claude' && source !== 'copilot') {
      res.status(400).json({ error: 'Bad request', message: 'Invalid source. Must be "claude" or "copilot"' });
      return;
    }

    const session = getSession(source as SessionSource, sessionId);

    if (!session) {
      res.status(404).json({ error: 'Not found', message: 'Session not found' });
      return;
    }

    res.json(session);
  } catch (error) {
    console.error('Error getting session:', error);
    res.status(500).json({ error: 'Internal server error', message: 'Failed to get session' });
  }
});

// GET /api/sessions/:source/:sessionId/messages - Get session messages with pagination
sessionsRouter.get('/:source/:sessionId/messages', (req, res) => {
  try {
    const { source, sessionId } = req.params;
    const offset = parseInt(req.query.offset as string) || 0;
    const limit = parseInt(req.query.limit as string) || 50;

    if (source !== 'claude' && source !== 'copilot') {
      res.status(400).json({ error: 'Bad request', message: 'Invalid source. Must be "claude" or "copilot"' });
      return;
    }

    const messages = getSessionMessages(source as SessionSource, sessionId, offset, limit);
    res.json(messages);
  } catch (error) {
    console.error('Error getting session messages:', error);
    res.status(500).json({ error: 'Internal server error', message: 'Failed to get messages' });
  }
});

// GET /api/sessions/:source/:sessionId/stats - Get session statistics
sessionsRouter.get('/:source/:sessionId/stats', (req, res) => {
  try {
    const { source, sessionId } = req.params;

    if (source !== 'claude' && source !== 'copilot') {
      res.status(400).json({ error: 'Bad request', message: 'Invalid source. Must be "claude" or "copilot"' });
      return;
    }

    const stats = getSessionStats(source as SessionSource, sessionId);

    if (!stats) {
      res.status(404).json({ error: 'Not found', message: 'Session not found' });
      return;
    }

    res.json(stats);
  } catch (error) {
    console.error('Error getting session stats:', error);
    res.status(500).json({ error: 'Internal server error', message: 'Failed to get stats' });
  }
});

// DELETE /api/sessions/:source/:sessionId - Delete a session
sessionsRouter.delete('/:source/:sessionId', (req, res) => {
  try {
    const { source, sessionId } = req.params;

    if (source !== 'claude' && source !== 'copilot') {
      res.status(400).json({ error: 'Bad request', message: 'Invalid source. Must be "claude" or "copilot"' });
      return;
    }

    // Find the session file
    const files = findSessionFiles(source as SessionSource);
    const filePath = files.get(sessionId);

    if (!filePath) {
      res.status(404).json({ error: 'Not found', message: 'Session not found' });
      return;
    }

    // Delete the file
    try {
      unlinkSync(filePath);
      console.log(`Deleted session file: ${filePath}`);
      res.json({ success: true, message: 'Session deleted successfully' });
    } catch (deleteError) {
      console.error(`Failed to delete session file ${filePath}:`, deleteError);
      res.status(500).json({ error: 'Internal server error', message: 'Failed to delete session file' });
    }
  } catch (error) {
    console.error('Error deleting session:', error);
    res.status(500).json({ error: 'Internal server error', message: 'Failed to delete session' });
  }
});

File: server-go/handlers/sessions.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"

	"agent-session-viewer/services"
	"agent-session-viewer/types"
)

// SessionsHandler binds all /api/sessions routes onto a chi.Router.
func SessionsHandler(r chi.Router) {
	r.Get("/", listSessions)
	r.Get("/{source}/{sessionId}", getSession)
	r.Get("/{source}/{sessionId}/messages", getMessages)
	r.Get("/{source}/{sessionId}/stats", getStats)
	r.Get("/{source}/{sessionId}/logfile", downloadLogFile)
	r.Delete("/{source}/{sessionId}", deleteSession)
}

func listSessions(w http.ResponseWriter, r *http.Request) {
	source := r.URL.Query().Get("source")
	sessions := services.ListSessions(source)
	if sessions == nil {
		sessions = []types.SessionSummary{}
	}
	jsonResponse(w, sessions)
}

func getSession(w http.ResponseWriter, r *http.Request) {
	src, ok := parseSource(w, chi.URLParam(r, "source"))
	if !ok {
		return
	}
	sessionID := chi.URLParam(r, "sessionId")
	detail := services.GetSession(src, sessionID)
	if detail == nil {
		http.Error(w, `{"error":"Not found"}`, http.StatusNotFound)
		return
	}
	jsonResponse(w, detail)
}

func getMessages(w http.ResponseWriter, r *http.Request) {
	src, ok := parseSource(w, chi.URLParam(r, "source"))
	if !ok {
		return
	}
	sessionID := chi.URLParam(r, "sessionId")
	offset, _ := strconv.Atoi(r.URL.Query().Get("offset"))
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	if limit <= 0 {
		limit = 50
	}
	msgs := services.GetSessionMessages(src, sessionID, offset, limit)
	if msgs == nil {
		msgs = []types.Message{}
	}
	jsonResponse(w, msgs)
}

func getStats(w http.ResponseWriter, r *http.Request) {
	src, ok := parseSource(w, chi.URLParam(r, "source"))
	if !ok {
		return
	}
	sessionID := chi.URLParam(r, "sessionId")
	stats := services.GetSessionStats(src, sessionID)
	if stats == nil {
		http.Error(w, `{"error":"Not found"}`, http.StatusNotFound)
		return
	}
	jsonResponse(w, stats)
}

func downloadLogFile(w http.ResponseWriter, r *http.Request) {
	src, ok := parseSource(w, chi.URLParam(r, "source"))
	if !ok {
		return
	}
	sessionID := chi.URLParam(r, "sessionId")
	files := services.FindSessionFiles(src)
	filePath, exists := files[sessionID]
	if !exists {
		http.Error(w, `{"error":"Not found"}`, http.StatusNotFound)
		return
	}
	if services.IsDBPath(filePath) {
		http.Error(w, `{"error":"No raw file","message":"Session is stored in SQLite. Use Export JSON instead."}`, http.StatusConflict)
		return
	}
	ext := ".log"
	if len(filePath) > 6 {
		ext = filePath[len(filePath)-6:]
		if ext[:1] != "." {
			ext = ".log"
		}
	}
	downloadName := fmt.Sprintf("%s-%s%s", src, sessionID, ext)
	w.Header().Set("Content-Disposition", `attachment; filename="`+downloadName+`"`)
	http.ServeFile(w, r, filePath)
}

func deleteSession(w http.ResponseWriter, r *http.Request) {
	src, ok := parseSource(w, chi.URLParam(r, "source"))
	if !ok {
		return
	}
	sessionID := chi.URLParam(r, "sessionId")
	if !services.DeleteSession(src, sessionID) {
		http.Error(w, `{"error":"Failed to delete session"}`, http.StatusInternalServerError)
		return
	}
	jsonResponse(w, map[string]interface{}{"success": true, "message": "Session deleted successfully"})
}

// ──────────────────────────────────────────────────────────
// Helpers
// ──────────────────────────────────────────────────────────

func parseSource(w http.ResponseWriter, raw string) (types.SessionSource, bool) {
	src := types.SessionSource(raw)
	if !src.IsValid() {
		http.Error(w, `{"error":"Bad request","message":"Invalid source"}`, http.StatusBadRequest)
		return "", false
	}
	return src, true
}

func jsonResponse(w http.ResponseWriter, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}
